lntemp/harness.go

```diff
@@ -80,3 +93,376 @@ func NewHarnessTest(t *testing.T, lndBinary string,
+			)
+			require.NoError(h, err)
+
+			addrScript, err := txscript.PayToAddrScript(addr)
+			require.NoError(h, err)
+
+			output := &wire.TxOut{
+				PkScript: addrScript,
+				Value:    10 * btcutil.SatoshiPerBitcoin,
+			}
+			_, err = h.Miner.SendOutputs(
+				[]*wire.TxOut{output}, 7500,
+			)
+			require.NoError(h, err, "send output failed")
+		}
+	}
+
+	// We generate several blocks in order to give the outputs created
+	// above a good number of confirmations.
+	h.Miner.MineBlocks(2)
+
+	// Now we want to wait for the nodes to catch up.
+	h.WaitForBlockchainSync(h.Alice)
+	h.WaitForBlockchainSync(h.Bob)
+
+	// Now block until both wallets have fully synced up.
+	expectedBalance := int64(btcutil.SatoshiPerBitcoin * 100)
+	err := wait.NoError(func() error {
+		aliceResp := h.Alice.RPC.WalletBalance()
+		bobResp := h.Bob.RPC.WalletBalance()
+
+		if aliceResp.ConfirmedBalance != expectedBalance {
+			return fmt.Errorf("expected 10 BTC, instead "+
+				"alice has %d", aliceResp.ConfirmedBalance)
+		}
+
+		if bobResp.ConfirmedBalance != expectedBalance {
+			return fmt.Errorf("expected 10 BTC, instead "+
+				"bob has %d", bobResp.ConfirmedBalance)
+		}
+
+		return nil
+	}, DefaultTimeout)
+	require.NoError(h, err, "timeout checking balance for node")
+}
+
+// Stop stops the test harness.
+func (h *HarnessTest) Stop() {
+	// Do nothing if it's not started.
+	if h.runCtx == nil {
+		h.Log("HarnessTest is not started")
+		return
+	}
+
+	// Stop all running nodes.
+	for _, node := range h.manager.activeNodes {
+		h.Shutdown(node)
+	}
+
+	close(h.lndErrorChan)
+
+	// Stop the fee service.
+	h.feeService.stop()
+
+	// Stop the chainBackend.
+	h.stopChainBackend()
+
+	// Stop the miner.
+	h.Miner.Stop()
+}
+
+// RunTestCase executes a harness test case. Any errors or panics will be
+// represented as fatal.
+func (h *HarnessTest) RunTestCase(testCase *TestCase) {
+	defer func() {
+		if err := recover(); err != nil {
+			description := errors.Wrap(err, 2).ErrorStack()
+			h.Fatalf("Failed: (%v) panic with: \n%v",
+				testCase.Name, description)
+		}
+	}()
+
+	testCase.TestFunc(h)
+}
+
+// resetStandbyNodes resets all standby nodes by attaching the new testing.T
+// and restarting them with the original config.
+func (h *HarnessTest) resetStandbyNodes(t *testing.T) {
+	for _, hn := range h.manager.standbyNodes {
+		// Inherit the testing.T.
+		h.T = t
+
+		// Reset the config so the node will be using the default
+		// config for the coming test. This will also inherit the
+		// test's running context.
+		h.RestartNodeWithExtraArgs(hn, hn.Cfg.OriginalExtraArgs)
+	}
+}
+
+// Subtest creates a child HarnessTest, which inherits the harness net and
+// stand by nodes created by the parent test. It will return a cleanup function
+// which resets  all the standby nodes' configs back to its original state and
+// create snapshots of each nodes' internal state.
+func (h *HarnessTest) Subtest(t *testing.T) (*HarnessTest, func()) {
```

Comment:
Great pattern, I suspect some of our issues/flakes are due to a test accumulating incremental unclean state from the ones that ran before it.